File: src/main/scala/List1.scala
object List1 extends App {

  def multiply(x:List[Double]):Double = {
    if (x.isEmpty) 0
    else if(x.length==1) x.head
    else x.head*multiply(x.tail)
  }

  def checkPositiveNumbers(x:List[Int]):Boolean ={
    if(x.isEmpty) true
    else if(x.head<0) false
    else checkPositiveNumbers(x.tail)
  }

  def stringBuilder(x:(List[String],String,String)):String={
    if(x._1.isEmpty) ""
    else if(x._1.length==1) x._1.head+x._3
    else x._1.head+x._2+stringBuilder(x._1.tail,x._2,x._3)
  }

  def factorial(x:Int):Int = {
    if(x<0) throw new IllegalArgumentException
    if(x==0) 1
    else x*factorial(x-1)
  }
}

File: src/main/ocaml/List1.ml
let rec multiply x =
  if x=[] then 0.
  else if (List.length x)=1 then List.hd x 
  else List.hd x *. multiply (List.tl x);;

  let myList = [1.;8.5;2.15];;
  print_float (multiply myList);;
  
  print_string "\n";;

let rec factorial x=
  if x<0 then raise(Failure "Ujemny argument")
  else if x=0 then 1
  else x*factorial(x-1);;


  print_string "\n";;
  print_int (factorial 3);;
  
  print_string "\n";;
  print_int (factorial 4);;
  
  print_string "\n";;
  print_int (factorial 6);;

  print_string "\n";;

let rec checkPositive x=
  if x=[] then true
  else if ((List.hd x)<0) then false
  else checkPositive (List.tl x);;
  
  print_string "\n";;
  let myList2= [1;2;3];;
  print_string (string_of_bool (checkPositive myList2));;
  print_string "\n";;

  
let rec stringBuilder x y z= 
  if x=[] then ""
  else if (List.length x)==1 then (List.hd x) ^ z
  else (List.hd x) ^ y ^ stringBuilder (List.tl x) y z;;

  print_string "\n";;
  let myList3=["Ala";"ma";"kota"];;
  print_string(stringBuilder myList3 " " "!");;
